clarity/contracts/arkadiko-stake-pool-diko-v2-1.clar

```diff
@@ -0,0 +1,487 @@
+
+      ;; Notify vesting
+      (try! (contract-call? vesting update-staking staker (get amount (get-stake-of staker))))
+
+      (ok amount)
+    )
+  )
+)
+
+;; @desc unstake tokens in the pool
+;; @param token; token to unstake
+;; @param amount; amount to unstake
+;; @param vesting; vesting contract to update with new stake amount
+;; @post uint; returns amount of tokens unstaked
+(define-public (unstake (token <ft-trait>) (amount uint) (vesting <vest-esdiko-trait>))
+  (let (
+    (staker tx-sender)
+  )
+    (asserts! (get-contract-active) ERR-INACTIVE)
+    (asserts! (or (is-eq (contract-of token) .arkadiko-token) (is-eq (contract-of token) .escrowed-diko-token)) ERR-WRONG-TOKEN)
+    (asserts! (or
+      (and (is-eq (contract-of token) .arkadiko-token) (>= (get diko (get-stake-of staker)) amount))
+      (and (is-eq (contract-of token) .escrowed-diko-token) (>= (get esdiko (get-stake-of staker)) amount))
+    ) ERR-INSUFFICIENT-STAKE)
+
+    ;; This method will increase the cumm-rewards-per-stake, and set it for the staker
+    ;; It also makes sure the multiplier points are credited to the staker
+    (try! (claim-pending-rewards))
+
+    (let (
+      ;; Calculate points to burn
+      (updated-stakes (get-stake-of staker))
+      (points-to-burn (/ (* (/ (* amount u1000000) (+ (get diko updated-stakes) (get esdiko updated-stakes))) (get points updated-stakes)) u1000000))
+    )
+      ;; Update total stake, increase cummulative rewards and update revenue
+      (var-set total-staked (- (var-get total-staked) amount points-to-burn))
+      (try! (update-revenue))
+      (try! (increase-cumm-reward-per-stake))
+
+      ;; Transfer tokens and update map
+      (try! (as-contract (contract-call? token transfer amount tx-sender staker none)))
+      (if (is-eq (contract-of token) .arkadiko-token)
+        (map-set stakes { staker: staker } (merge updated-stakes { 
+          amount: (- (get amount updated-stakes) amount points-to-burn),
+          diko: (- (get diko updated-stakes) amount),
+          points: (- (get points updated-stakes) points-to-burn)
+        }))
+        (map-set stakes { staker: staker } (merge updated-stakes { 
+          amount: (- (get amount updated-stakes) amount points-to-burn),
+          esdiko: (- (get esdiko updated-stakes) amount),
+          points: (- (get points updated-stakes) points-to-burn)
+        }))
+      )
+
+      ;; Update user cumm-reward-per-stake
+      (map-set stake-rewards { staker: staker, token: .usda-token } { cumm-reward-per-stake: (get cumm-reward-per-stake (get-reward-of .usda-token)) })
+      (map-set stake-rewards { staker: staker, token: .escrowed-diko-token } { cumm-reward-per-stake: (get cumm-reward-per-stake (get-reward-of .escrowed-diko-token)) })
+
+      ;; Notify vesting
+      (try! (contract-call? vesting update-staking staker (get amount (get-stake-of staker))))
```

Comment:
Q: The vesting contract is called by the sender, not by the contact. So I don't think it can do any harm.
In the contract itself we check using `contract-caller` if it's being called by the pool contract.